### PR TITLE
Add the offline removal, BCD and nested-VM helpers (3 of 3)

### PR DESCRIPTION
**Pull request by mvaferreira**
## What this adds

The three helpers that perform destructive or boot-affecting repairs.

| Helper | Used by | Purpose |
|---|---|---|
| `Use-OfflineFileRemoval.ps1` | 2 | Removes a set of files from an offline disk with a verified backup, a proof that only the intended files went, and automatic rollback when that proof fails |
| `Get-OfflineBcdStore.ps1` | 2 | Reads and modifies the Boot Configuration Data store of an offline installation |
| `Use-NestedRepairVm.ps1` | 1 | Drives the nested Hyper-V guest that `az vm repair create --enable-nested` provides |

No `map.json` entries are added. Helpers are not run-ids, so nothing new is invocable yet.

## Part of the #143 split

This came from the three-way split of #143. **#143 is now merged and stays unchanged.**
#146 carries the additional shared state/discovery/coordination support used here. The briefly
opened #148 split is closed; all remaining fixes stay in **#146 and #147**, which merge cleanly
in either order. Consuming scenario PRs must wait for both.

## Command injection in the BCD helper

`Invoke-BcdEdit` built a command string and ran it through `cmd.exe /c`. Any value that reached it
from the offline store — a description field, a device path — was re-parsed by the shell.

It now takes `[string[]]$Arguments` and invokes `& bcdedit.exe /store $StorePath @Arguments`
directly. No shell, no re-parsing. This was verified rather than assumed: under the old shape a
payload embedded in an argument executed; under the new shape the identical payload arrives as a
single literal argument.

The signature change touches exactly one scenario, `win-fix-bcd.ps1`, at three call sites, all of
which already build proper argument arrays.

## Removing files from an offline disk

The review found that the "two-layer" allow-list was extension-only on both layers. In the
documented `System32\config` use case that made `SYSTEM.LOG1`, `SOFTWARE.LOG2` and the transaction
`.blf` files removable — deleting a hive's log alongside the hive is how a recoverable image becomes
an unbootable one.

There is now a **base-name veto**: a file whose name without its extension matches a hive name is
refused, ordinal and case-insensitive. This was checked against the real scenarios rather than
assumed safe — genuine transaction files carry a GUID and a `.TM`/`.TxR` infix, so their base name
never equals a hive name and `win-fix-transaction-logs` is unaffected.

Rollback was also silently ineffective: it enumerated the backup with `-ErrorAction
SilentlyContinue`, so a rollback that recovered nothing reported `Restored = 0; Failed = 0` and read
as success. It now fails loudly when the recovered count does not match, routes restores through the
protected-file path so an ACL cannot block them, and reports a distinct fatal outcome when rollback
itself fails.

The initial September 10 dependency follow-up updated hive-comparison documentation to match #146:
`Test-OfflineHiveFile` now parses through the shared in-memory offreg reader from #143, not a
scratch registry mount. The before/after `Loads` field remains a comparison result, not a promise
that Windows will boot. The subsequent review fixes below change removal, rollback and VM coordination.

## Capacity, hashes and rollback

`Get-OfflineFreeSpace` now uses `GetDiskFreeSpaceExW` on the actual directory's volume, without
depending on the PowerShell drive list. Real volume-GUID and boot-volume paths are exercised.
Unknown, invalid or insufficient capacity refuses removal before any backup copy or deletion;
zero bytes remains a valid capacity result, not an unknown one.

The existing backup comparison already rejected missing hashes, and
`Test-OfflineRemovalResult` did not contain a null-equals-null hash comparison. The gap was later
restoration: verified original hashes now survive in `BackupRecord`, and rollback checks both
the backup before copying and the destination afterward. A missing comparison hash is a failure.
Consumers must persist these records and pass them as `-FileRecord` on later revert.

Native testing also found that applying an ACL after attributes set the NTFS Archive bit.
Rollback now restores security first, then reapplies and reads back exact attributes. Failed
metadata replay is a failed restore, not a warning alongside `Succeeded = true`. Legacy records
without a Hash field explicitly warn that they verify the copy against the current backup,
not against a recorded historical original.

## The nested guest

`Start-NestedRepairVm` passed the wrong variable to the attach call, so the disk it took offline was
not the disk it handed to the guest. Four early returns left host disks offline on failure. An
already-running guest short-circuited to `Started = $true` without checking whether the requested
disk was attached at all.

All three are fixed, and a fourth case that the review did not raise is now refused at both ends: if
every requested disk is skipped, the guest would previously start with **no disk**, report success,
and leave the caller waiting out the full heartbeat timeout.

Disks are restored in a `finally`, and only the disks this call took offline — a disk that was
already offline at entry was offline for a reason this helper does not own.

The new coordination fix marks started/adopted guests in Notes before reporting success.
The shared support in #146 recognizes that marker across processes and refuses discovery while
the guest is active. It also protects unrelated guests and serializes lifecycle transitions.
The temporary-user caller no longer invokes the broad stopper after its explicit shutdown;
it confirms `Stopped` before rediscovery. An unconfirmed/failed forced stop is not success.

## First consumer and end-to-end scope

The first planned upstream consumer of `Invoke-OfflineRemovalPlan` is
[`win-fix-transaction-logs`](https://github.com/mvaferreira/repair-script-library/blob/main/src/windows/win-fix-transaction-logs.ps1)
in wave 1. `win-fix-pending-servicing` also consumes it, in wave 3.

This review follow-up exercised real attached **MBR/Gen1 and GPT/Gen2 VHD volumes**, including a
FAT32 EFI partition: native capacity checks, protected file deletion/restoration, same-size
backup tampering, original ACL/attribute restoration, automatic rollback, and BCD
create/edit/read-back/restore through each generation-specific path. **51 assertions passed.**
The originals were preserved and the fixtures removed.

An additional **22 native assertions** exercised the real manifest writers and both scenario
revert entry points in four separate PowerShell processes. Each refused a same-size modified
backup with `STATUS_ERROR`, retained its manifest, and completed a verified retry after the
original backup was restored. Pending servicing also kept the pending.xml rename retryable.
These used synthetic files/hives on a disposable attached VHD, not a complete Windows repair.

These are full helper-level operations on real volumes, not bootable customer-image fixtures.
They do **not** establish a new full Gen1/Gen2 Windows scenario
`az vm repair create/run --preview/restore` acceptance matrix. That distinction is retained
explicitly; scenario acceptance belongs to the consuming PRs.

## Conventions followed

- Every repair is evidence-driven: nothing is written unless a specific fault is detected, so a
  healthy image produces no changes.
- No security protection is disabled as a workaround.
- Logging is ordered so the verdict is last, because `az vm run-command` keeps only the final 4096
  characters of the output stream.

## Testing

- **PSScriptAnalyzer: 0 findings** on all three files, at every severity.
- **66/66 removal regressions** and **87/87 nested lifecycle regressions** pass on PS5.1 and PS7.
- **57/57 consumer regressions** pass on both engines: verified records survive JSON/merged
  manifests, failed or partial restores keep retry data and return error, and successful retries
  complete without misleading rollback claims.
- Defect-level tests covering bcdedit argument injection, the hive base-name veto (including real
  transaction-log filenames that must stay removable), the nested-VM disk hand-off, and `-WhatIf`
  safety on every function that changes host state.
- No `map.json` run-id resolves to a helper, so these files cannot be exercised through
  `az vm repair run` on their own. They are validated by the test suite here; the scenario scripts
  that consume them are validated end-to-end in their own pull requests. `win-fix-bcd.ps1` is the
  one existing script whose call sites change (`Invoke-BcdEdit -Command` → `-Arguments`, 3 sites),
  and it ships in its own pull request with that validation.
- Earlier nested Hyper-V lab validation is recorded in the
  [September 8 follow-up](https://github.com/Azure/repair-script-library/pull/143#issuecomment-5589235133).
- This round adds **93 native Hyper-V assertions** with disposable Gen1/Gen2 guests and actual
  passthrough-disk hand-offs. A separate PowerShell process cannot interrupt a helper-owned guest.
  Five existing guests and seven original disks were preserved. These guests have no Windows OS;
  their role is lifecycle/disk coordination, not boot or payload validation.

### CODE REVIEW

File: src/windows/common/helpers/Get-OfflineBcdStore.ps1 [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+    #>
+    param(
+        [Parameter(Mandatory = $true)][string]$StorePath,
+        [Parameter(Mandatory = $true)][ValidateNotNullOrEmpty()][string[]]$Arguments
+    )
+
+    $raw = & bcdedit.exe /store $StorePath @Arguments 2>&1
+    $exitCode = $LASTEXITCODE
+
+    return [PSCustomObject]@{
```
Comment (EdwinBernal1) [inline src/windows/common/helpers/Get-OfflineBcdStore.ps1:132]: ### Warning — `Get-OfflineFreeSpace` returns `$null` on exactly the drives this library creates

**`src/windows/common/helpers/Use-OfflineFileRemoval.ps1` L122–L136**:

```powershell
L132    $drive = Get-PSDrive -Name $root.Substring(0, 1) -ErrorAction Stop
L133    return [int64]$drive.Free
L135    catch { return $null }
```

`OfflineRepairCommon.ps1` opens with the reason this does not work:

> *Join-Path and Test-Path throw DriveNotFoundException when a path refers to a drive letter that is not a live PowerShell drive. Offline repairs work with letters that come and go...*

That is why `Join-OfflinePath` and `Test-OfflinePath` exist. But **L132** reaches for `Get-PSDrive`, which resolves against the same PowerShell drive list — and the letters `Add-PartitionDriveLetter` (#143 `Get-OfflineWindowsDisk.ps1` **L402**) assigns to hidden EFI System and Recovery partitions via `diskpart` are precisely the ones that are not reliably there. `Get-PSDrive` throws, the `catch` at **L135** returns `$null`, and at **L889–L891** the consumer does this:

```powershell
L887    $required = ($Plan.TotalBytes * 2)
L888    $free = Get-OfflineFreeSpace -Path $backupPath
L889    if ($null -eq $free) {
L890        Add-OfflineRepairLog -Level Warning -Message 'Free space on the backup volume could not be confirmed; continuing.'
L891    }
```

So the guard is weakest exactly where the risk is highest: writing a backup to a small EFI or Recovery partition before deleting from it. If the copy then fails on a full volume, the removal plan has a backup it cannot restore from.

**Suggested fix — query the volume, not the PowerShell drive** (replacing L130–L134):

```powershell
$letter = $root.Substring(0, 1)
$volume = Get-Volume -DriveLetter $letter -ErrorAction SilentlyContinue
if ($volume) { return [int64]$volume.SizeRemaining }

# Falls through for hidden partitions Get-Volume does not surface by letter.
$logicalDisk = Get-CimInstance Win32_LogicalDisk -Filter "DeviceID='${letter}:'" -ErrorAction SilentlyContinue
if ($logicalDisk) { return [int64]$logicalDisk.FreeSpace }
return $null
```

Separately, **L889–L891**: please make `$null` **refuse the removal** rather than warn and continue. "I could not determine free space" should not be treated the same as "there is enough" — that is the distinction `Get-OfflineHiveKeyState` in #146 makes correctly between `Absent` and `Unknown`, and it should hold here too.

---

File: src/windows/common/helpers/Use-OfflineFileRemoval.ps1 [hunk 2]
```diff
@@ -0,0 +1,1025 @@
...
+        throw "Use-OfflineFileRemoval.ps1 requires '$required', which its dependencies did not define."
+    }
+}
+
+function Get-OfflineFileHashValue {
+    <#
+    .SYNOPSIS
+        SHA256 of a file, or $null when it cannot be read.
+    #>
+    param([Parameter(Mandatory = $true)][string]$Path)
```
Comment (EdwinBernal1) [inline src/windows/common/helpers/Use-OfflineFileRemoval.ps1:127]: ### Warning — `Get-OfflineFileHashValue` swallows the failure it exists to detect

**L112–L119**:

```powershell
L117    try { return (Get-FileHash -LiteralPath $Path -Algorithm SHA256 -ErrorAction Stop).Hash }
L118    catch { return $null }
```

The hash exists so a rollback can prove the restored file is byte-identical. If hashing fails on the way in, `$null` is stored; if it fails on the way out, `$null` is compared. `$null -eq $null` is `$true`, so a verification that never actually ran can read as a successful match, depending on how `Test-OfflineRemovalResult` (**L654**) compares them.

I have not traced every comparison path, so this may already be handled — if so, ignore me. If it is not, the fix is to treat a `$null` hash as **unverifiable** and say so, rather than letting it flow into an equality test.

This is the same lesson `Test-OfflineFileSignature` in #143 already applies correctly, and its help says it plainly: *"A caller must not treat None as either trusted or untrusted — it means the file could not be checked."* That principle should reach the hashing path too.

---

File: src/windows/common/helpers/Use-NestedRepairVm.ps1 [hunk 3]
```diff
@@ -0,0 +1,902 @@
...
+        catch {
+            $result.Reason = "the boot order of nested guest '$($Vm.Name)' could not be set, so it cannot be made to boot from its disk: $($_.Exception.Message)"
+            return $result
+        }
+    }
+
+    return $result
+}
+
+function Start-NestedRepairVm {
```
Comment (EdwinBernal1) [inline src/windows/common/helpers/Use-NestedRepairVm.ps1:433]: ### Question — `Use-NestedRepairVm` and `Stop-NestedRepairVm`

`Get-OfflineWindowsDisk` (#143 **L266–L303**) calls `Stop-NestedRepairVm`, which turns off **every running Hyper-V VM** on the rescue VM:

```powershell
#143 L282   $running = @(Get-VM -ErrorAction SilentlyContinue | Where-Object { $_.State -eq 'Running' })
#143 L299   Stop-VM -Name $vm.Name -TurnOff -Force -ErrorAction SilentlyContinue
```

`-TurnOff` is the power-cut equivalent, not a graceful shutdown. On a rescue VM created with `--enable-nested` that is the intent and it is fine. But this PR adds `Start-NestedRepairVm` (`Use-NestedRepairVm.ps1` **L381**) and `Stop-NestedRepairVmGraceful` (**L765**) — so once both are in, does a discovery pass from one script hard-power-off a nested repair VM another part of the flow deliberately started?

If the two are always sequential this is a non-issue. If they can overlap, `Stop-NestedRepairVm` should skip VMs it recognises as managed by `Use-NestedRepairVm` (a name prefix or a note field would do), or prefer `Stop-NestedRepairVmGraceful` for those.

---

**Review by EdwinBernal1 — CHANGES_REQUESTED** [review]
### Info — scope and testing

Same structural point as #143: nine helpers, no consuming script, no tests, and `tests/` is still a placeholder. For read-mostly helpers that is a manageable risk. For **file removal and BCD editing on a customer's boot volume** it is a larger one — these are the helpers where a defect is unrecoverable for the customer, and they are landing with no in-repo exercise at all.

Not asking for a test harness in this PR. But it would help reviewers a lot if the description said which scenario will consume `Invoke-OfflineRemovalPlan` (`Use-OfflineFileRemoval.ps1` **L785**) first, and whether it has been run end-to-end against a real attached disk (Gen1 and Gen2, since `Get-BcdStorePath` at `Get-OfflineBcdStore.ps1` **L139** branches on generation).

---

File: src/windows/common/helpers/Get-OfflineBcdStore.ps1 (hunk 1 above)
Comment (mvaferreira) [reply src/windows/common/helpers/Get-OfflineBcdStore.ps1:132]: Addressed in `Use-OfflineFileRemoval.ps1`. Capacity is now queried with `GetDiskFreeSpaceExW` against the actual directory's volume, independently of PowerShell drive registration. A genuine zero-byte result remains distinct from an unknown result. Unknown/invalid capacity, arithmetic overflow and insufficient space all refuse before any backup copy or deletion.

Native coverage includes volume-GUID paths and real MBR/Gen1 and GPT/Gen2 volumes, including a FAT32 EFI partition. The removal regression suite passes 66/66 on PS5.1 and PS7; the native removal/BCD suite passes 51 assertions.

File: src/windows/common/helpers/Use-OfflineFileRemoval.ps1 (hunk 2 above)
Comment (mvaferreira) [reply src/windows/common/helpers/Use-OfflineFileRemoval.ps1:127]: I traced the comparison paths: the original backup checks already rejected missing hashes, and `Test-OfflineRemovalResult` did not contain a null-equals-null hash comparison. There was a later-restore gap worth closing, though. Verified original hashes now survive in `BackupRecord`; rollback verifies the backup before copying and the restored destination afterward. Missing/unreadable comparison hashes fail rather than matching.

Both downstream consumers now persist the records and pass `-FileRecord` on revert. Failed or partial restores return error and retain the manifest. A legacy record with no Hash field explicitly warns that it can verify only against the current backup; an explicitly empty Hash fails.

Native testing also exposed an attribute-order issue: ACL replay could add Archive after attributes had been restored. Security is now replayed first, exact attributes last with read-back, and metadata replay failure makes the restore fail. The live run confirmed same-size backup tampering is refused. An additional 22 native assertions through four actual scenario revert processes confirmed error/manifest retention and successful verified retries for both consumers.

File: src/windows/common/helpers/Use-NestedRepairVm.ps1 (hunk 3 above)
Comment (mvaferreira) [reply src/windows/common/helpers/Use-NestedRepairVm.ps1:433]: Addressed for overlapping flows, not just sequential ones. Started/adopted guests receive a persistent exact Notes marker before success, and shared discovery support in #146 recognizes it across processes. Discovery refuses an active managed guest rather than merely skipping its shutdown and then onlining its attached disk. It also refuses unrelated active guests; automatic shutdown is restricted to the selected legacy ProblemVM or explicit VM Id.

A shared reentrant host mutex serializes lifecycle transitions. Explicit stop must confirm Off; an unconfirmed forced stop is not success. The temporary-user caller now confirms `Stopped` before rediscovery and no longer invokes a redundant broad stopper.

The dedicated suite passes 87/87 on PS5.1 and PS7. Native Hyper-V coverage passed 93 assertions using disposable Gen1/Gen2 guests, passthrough disks and a separate discovery process, while preserving all five existing guests and seven original disks. These guests were firmware-only fixtures, not Windows payload/boot tests.

**Comment by mvaferreira** [comment]
Updated the description and answered all three inline threads. The first planned upstream consumer of `Invoke-OfflineRemovalPlan` is `win-fix-transaction-logs` (wave 1); `win-fix-pending-servicing` also consumes it (wave 3). Their updated record/restore handling is in the fork, with 57/57 consumer regressions passing on PS5.1 and PS7.

For the generation question: this round ran 51 native assertions on real attached MBR/Gen1 and GPT/Gen2 volumes, including FAT32 EFI, covering capacity, protected removal/restore, tamper refusal, exact ACL/attribute replay, automatic rollback and BCD create/edit/read-back/restore through both generation-specific paths. Another 22 assertions ran both actual scenario revert entry points in four processes against synthetic attached-volume fixtures. The native Hyper-V lifecycle run passed 93 assertions.

These are real native helper/entry-point operations, but the hives/BCD stores were synthetic and the nested guests had no Windows OS. They are not a fresh full Gen1/Gen2 `az vm repair create/run --preview/restore` scenario acceptance matrix; that remains explicitly scoped to the consuming PRs. Private lab/test tooling is not being added to this helper PR.

The remaining work stays in #146 and #147 only. #143 is already merged and unchanged; #146 carries the shared support used here; #148 is closed. Both remaining merge orders are clean, and scenario PRs wait for both. Threads are left open for reviewer confirmation.

**Review by EdwinBernal1 — APPROVED** [review]
